contracts/extensions/MultisigPermissions.sol

```diff
@@ -0,0 +1,640 @@
+  ) public {
+    // This is not strictly necessary, since these roles are never used in subdomains
+    require(_roles & ROOT_ROLES == 0 || _domainId == 1, "multisig-bad-domain-for-role");
+
+    require(
+      validateDomainInheritance(_permissionDomainId, _childSkillIndex, _domainId),
+      "multisig-invalid-domain-inheritance"
+    );
+
+    // Allow this function to be called if the caller:
+    require(
+      // Has core root permissions OR
+      colony.hasUserRole(msgSender(), 1, ColonyDataTypes.ColonyRole.Root) ||
+        // Has core architecture, if we're using that permission in a child domain
+        (colony.hasUserRole(
+          msgSender(),
+          _permissionDomainId,
+          ColonyDataTypes.ColonyRole.Architecture
+        ) && (_permissionDomainId != _domainId)),
+      "multisig-caller-not-correct-permissions"
+    );
+
+    Domain memory domain = colony.getDomain(_domainId);
+
+    bytes32 existingRoles = getUserRoles(_user, _domainId);
+    bytes32 rolesChanged = _roles ^ existingRoles;
+    bytes32 roles = _roles;
+    bool setTo;
+
+    for (uint8 roleId; roleId < uint8(ColonyRole.NUMBER_OF_ROLES); roleId += 1) {
+      bool changed = uint256(rolesChanged) % 2 == 1;
+      if (changed) {
+        setTo = uint256(roles) % 2 == 1;
+        setUserRole(_user, _domainId, roleId, setTo);
+
+        if (setTo) {
+          domainSkillRoleCounts[domain.skillId][roleId] += 1;
+        } else {
+          domainSkillRoleCounts[domain.skillId][roleId] -= 1;
+        }
+
+        emit MultisigRoleSet(msgSender(), _user, _domainId, roleId, setTo);
+      }
+
+      roles >>= 1;
+      rolesChanged >>= 1;
+    }
+  }
+
+  // Internal functions
+
+  function getSig(bytes memory motion) internal pure returns (bytes4 sig) {
+    assembly {
+      sig := mload(add(motion, 0x20))
+    }
+  }
+
+  function validateDomainInheritance(
+    uint256 _permissionDomainId,
+    uint256 _childSkillIndex,
+    uint256 _domainId
+  ) internal view returns (bool) {
+    uint256 colonyDomainCount = colony.getDomainCount();
+    require(
+      _permissionDomainId > 0 && _permissionDomainId <= colonyDomainCount,
+      "multisig-domain-does-not-exist"
+    );
+    require(_domainId > 0 && _domainId <= colonyDomainCount, "multisig-domain-does-not-exist");
+
+    if (_permissionDomainId == _domainId) {
+      return _childSkillIndex == UINT256_MAX;
+    } else {
+      Domain memory domain = colony.getDomain(_permissionDomainId);
+      Domain memory childDomain = colony.getDomain(_domainId);
+      uint256 childSkillId = colonyNetwork.getChildSkillId(domain.skillId, _childSkillIndex);
+      return childSkillId == childDomain.skillId;
+    }
+  }
+
+  function validateMotionDomain(
+    uint256 _permissionDomainId,
+    uint256 _childSkillIndex,
+    uint256 _motionId
+  ) internal view {
+    uint256 permissionSkillId = colony.getDomain(_permissionDomainId).skillId;
+    uint256 userActingInDomainWithSkillId = colonyNetwork.getChildSkillId(
```

Comment:
What happens with `_childSkillIndex = UINT256_MAX` here?

Reply:
We hit [the branch](https://github.com/JoinColony/colonyNetwork/blob/485a950b81101253fd5614598e5a0bc1e6842d6b/contracts/colonyNetwork/ColonyNetwork.sol#L165) for that special case.

(I assume this comment is about whether it works at all, rather than whether it should be short-circuited here).